File: core/hooks/policy_engine.py
"""
Universal Agentic Hooks Framework (UAHF) — Policy Engine
========================================================
Centralized, deterministic policy evaluation pipeline for governing agent behavior.
Integrates safety hooks, secret sanitizers, package policy, and TDD guards.
"""

from pathlib import Path
import re
import sys
from typing import Any, Dict, List, Optional, Set

from .types import (
    HookEvent,
    HookResult,
    HookSource,
    HookType,
    HookVerdict,
    PolicyRule,
)


class DestructiveCommandRule(PolicyRule):
    """Blocks catastrophic commands (rm -rf /, dd if=, mkfs, git -f, curl | sh)."""
    rule_id = "SEC-001-DESTRUCTIVE-COMMAND"
    description = "Blocks destructive system, git, and exfiltration commands"

    NETWORK_PATTERNS = [
        (re.compile(r"\bcurl\b.*\|\s*(ba)?sh\b"), "curl piped to shell is blocked. Download, inspect, and execute manually."),
        (re.compile(r"\bwget\b.*\|\s*(ba)?sh\b"), "wget piped to shell is blocked. Download, inspect, and execute manually."),
    ]

    SYSTEM_PATTERNS = [
        (re.compile(r"\bdd\b\s+if="), "dd command with raw input file is blocked. Irreversible disk write risk."),
        (re.compile(r"\bmkfs\b"), "mkfs command is blocked. Filesystem formatting destroys all data."),
    ]

    GIT_PATTERNS = [
        (
            re.compile(r"\bgit\s+push\b.*(?<![-\w])--force(?![-\w])"),
            "git push --force is blocked. Use --force-with-lease to protect remote history.",
        ),
        (
            re.compile(r"\bgit\s+push\b.*\s-[a-zA-Z]*f"),
            "git push -f is blocked. Use --force-with-lease to protect remote history.",
        ),
        (re.compile(r"\bgit\s+reset\b.*\s--hard\b"), "git reset --hard is blocked. Discards uncommitted work permanently."),
        (re.compile(r"\bgit\s+checkout\b\s+(?:--\s+)?\."), "git checkout . is blocked. Discards unstaged modifications."),
        (re.compile(r"\bgit\s+restore\b\s+\."), "git restore . is blocked. Discards unstaged modifications."),
        (re.compile(r"\bgit\s+clean\b.*\s-[a-zA-Z]*f"), "git clean -f is blocked. Permanently deletes untracked files."),
        (re.compile(r"\bgit\s+branch\b.*\s-D\b"), "git branch -D is blocked. Use git branch -d for safe deletion."),
        (
            re.compile(r"\bgit\s+branch\b.*\s--delete\b.*\s--force\b"),
            "git branch --delete --force is blocked. Use git branch -d for safe deletion.",
        ),
    ]

    DANGEROUS_TARGETS = {"/", "/*", "~", "~/", "$HOME", "$HOME/", "$HOME/*"}

    def _check_dangerous_rm(self, sub_command: str) -> Optional[str]:
        tokens = sub_command.split()
        if not tokens or tokens[0] != "rm":
            return None

        flags = ""
        targets = []
        for token in tokens[1:]:
            if token.startswith("-") and not token.startswith("--"):
                flags += token[1:]
            elif not token.startswith("-"):
                targets.append(token.strip("\"'"))

        has_recursive = "r" in flags or "R" in flags
        has_force = "f" in flags

        if not (has_recursive and has_force):
            return None

        for target in targets:
            if target in self.DANGEROUS_TARGETS:
                return f"rm -rf targeting {target} is blocked. Catastrophic, irreversible file deletion."
        return None

    def evaluate(self, event: HookEvent) -> Optional[HookResult]:
        command = event.command or (event.args.get("command") if isinstance(event.args, dict) else None)
        if not command or not isinstance(command, str):
            return None

        for pattern, msg in self.NETWORK_PATTERNS:
            if pattern.search(command):
                return HookResult(
                    event_id=event.event_id,
                    verdict=HookVerdict.BLOCK,
                    message=f"DESTRUCTIVE COMMAND BLOCKED: {msg}",
                    exit_code=2,
                    rule_id=self.rule_id,
                )

        for pattern, msg in self.SYSTEM_PATTERNS:
            if pattern.search(command):
                return HookResult(
                    event_id=event.event_id,
                    verdict=HookVerdict.BLOCK,
                    message=f"DESTRUCTIVE COMMAND BLOCKED: {msg}",
                    exit_code=2,
                    rule_id=self.rule_id,
                )

        for pattern, msg in self.GIT_PATTERNS:
            if pattern.search(command):
                return HookResult(
                    event_id=event.event_id,
                    verdict=HookVerdict.BLOCK,
                    message=f"DESTRUCTIVE COMMAND BLOCKED: {msg}",
                    exit_code=2,
                    rule_id=self.rule_id,
                )

        for sub_cmd in re.split(r"\s*(?:&&|\|\||;)\s*", command):
            for segment in sub_cmd.split("|"):
                rm_err = self._check_dangerous_rm(segment.strip())
                if rm_err:
                    return HookResult(
                        event_id=event.event_id,
                        verdict=HookVerdict.BLOCK,
                        message=f"DESTRUCTIVE COMMAND BLOCKED: {rm_err}",
                        exit_code=2,
                        rule_id=self.rule_id,
                    )

        return None


class PackagePolicyRule(PolicyRule):
    """Enforces package manager hygiene: bans Colima, restricts sudo brew, verifies safe installs."""
    rule_id = "PKG-002-PACKAGE-HYGIENE"
    description = "Enforces container and package manager governance policies"

    FORBIDDEN_PACKAGES = {
        "colima": "Colima is prohibited per project constitution due to large footprint. Use lightweight alternatives.",
    }

    BREW_INSTALL_REGEX = re.compile(r"\bbrew\s+(?:install|reinstall|cask)\s+([^\s;]+)", re.IGNORECASE)
    SUDO_BREW_REGEX = re.compile(r"\bsudo\s+brew\b", re.IGNORECASE)

    def evaluate(self, event: HookEvent) -> Optional[HookResult]:
        command = event.command or (event.args.get("command") if isinstance(event.args, dict) else None)
        args_target = event.args.get("target") or event.args.get("package")

        if command and isinstance(command, str):
            if self.SUDO_BREW_REGEX.search(command):
                return HookResult(
                    event_id=event.event_id,
                    verdict=HookVerdict.BLOCK,
                    message="PACKAGE POLICY VIOLATION: 'sudo brew' is prohibited. Homebrew must not run as root.",
                    exit_code=2,
                    rule_id=self.rule_id,
                )

            match = self.BREW_INSTALL_REGEX.search(command)
            if match:
                pkg_candidate = match.group(1).lower().strip("\"'")
                for forbidden_pkg, reason in self.FORBIDDEN_PACKAGES.items():
                    if forbidden_pkg in pkg_candidate:
                        return HookResult(
                            event_id=event.event_id,
                            verdict=HookVerdict.BLOCK,
                            message=f"PACKAGE POLICY VIOLATION: Package '{forbidden_pkg}' is blocked. {reason}",
                            exit_code=2,
                            rule_id=self.rule_id,
                        )

        if args_target and isinstance(args_target, str):
            tgt = args_target.lower()
            for forbidden_pkg, reason in self.FORBIDDEN_PACKAGES.items():
                if forbidden_pkg == tgt or forbidden_pkg in tgt:
                    return HookResult(
                        event_id=event.event_id,
                        verdict=HookVerdict.BLOCK,
                        message=f"PACKAGE POLICY VIOLATION: Package '{forbidden_pkg}' is blocked. {reason}",
                        exit_code=2,
                        rule_id=self.rule_id,
                    )

        return None


class SensitiveFileRule(PolicyRule):
    """Prevents overwriting or exposing security-sensitive files (.env, keys, certs)."""
    rule_id = "SEC-003-SENSITIVE-FILES"
    description = "Prevents tampering with credentials, environment secrets, and private keys"

    SENSITIVE_PATTERNS = [
        (re.compile(r"(^|[/\\])\.env(\.[a-zA-Z0-9_-]+)?$"), "Environment secret file (.env)"),
        (re.compile(r"\.(pem|key|p12|pfx)$", re.IGNORECASE), "Private key or certificate"),
        (re.compile(r"(^|[/\\])id_(rsa|ed25519|ecdsa|dsa)$"), "SSH private key"),
        (re.compile(r"(^|[/\\])(credentials|secrets|passwords)\.(json|ya?ml|toml)$", re.IGNORECASE), "Secrets store"),
        (re.compile(r"(^|[/\\])(service[-_]?account|token)\.json$", re.IGNORECASE), "Service account / Token file"),
        (re.compile(r"\.(tfstate|tfvars)$", re.IGNORECASE), "Terraform state / variable secrets"),
    ]

    def evaluate(self, event: HookEvent) -> Optional[HookResult]:
        file_path = event.file_path or event.args.get("file_path") or event.args.get("path")
        if not file_path or not isinstance(file_path, str):
            return None

        for pattern, label in self.SENSITIVE_PATTERNS:
            if pattern.search(file_path):
                # If tool attempts write/edit on sensitive file, block or warn
                tool_name = (event.tool_name or "").lower()
                is_write = (
                    "write" in tool_name
                    or "edit" in tool_name
                    or event.hook_type in [HookType.PRE_TOOL, HookType.PRE_COMMAND]
                )
                if is_write:
                    return HookResult(
                        event_id=event.event_id,
                        verdict=HookVerdict.WARN,
                        message=f"SECURITY ADVISORY: Access/modification to sensitive file '{file_path}' ({label}).",
                        exit_code=0,
                        rule_id=self.rule_id,
                    )
        return None


class TddIntegrityRule(PolicyRule):
    """Enforces TDD discipline by blocking test file edits when .tdd-guard is active."""
    rule_id = "GOV-004-TDD-INTEGRITY"
    description = "Protects test files from modification during implementation phases"

    TEST_FILE_PATTERNS = [
        re.compile(r"(^|[/\\])test_[^/\\]+\.py$"),
        re.compile(r"[._]test\.[jt]sx?$"),
        re.compile(r"[._]spec\.[jt]sx?$"),
    ]

    def __init__(self, project_dir: Optional[str] = None):
        self.project_dir = Path(project_dir) if project_dir else Path.cwd()

    def is_guard_active(self) -> bool:
        return (self.project_dir / ".tdd-guard").exists() or Path(".tdd-guard").exists()

    def evaluate(self, event: HookEvent) -> Optional[HookResult]:
        if not self.is_guard_active():
            return None

        file_path = event.file_path or event.args.get("file_path") or event.args.get("path")
        if not file_path or not isinstance(file_path, str):
            return None

        tool_name = (event.tool_name or "").lower()
        if not ("edit" in tool_name or "write" in tool_name):
            return None

        for pattern in self.TEST_FILE_PATTERNS:
            if pattern.search(file_path):
                return HookResult(
                    event_id=event.event_id,
                    verdict=HookVerdict.BLOCK,
                    message=(
                        f"TDD GUARD ACTIVE: Modification of test file '{file_path}' is blocked. "
                        "Modify implementation code to make tests pass."
                    ),
                    exit_code=2,
                    rule_id=self.rule_id,
                )
        return None


class SecretLeakRule(PolicyRule):
    """Scans tool outputs and command responses for accidental credential exfiltration."""
    rule_id = "SEC-005-SECRET-LEAK-FILTER"
    description = "Detects and blocks leakage of API keys and authentication tokens in telemetry"

    PATTERNS = [
        (re.compile(r"sk-(?:proj|ant|live)-[a-zA-Z0-9_\-]{20,}"), "OpenAI / Anthropic API Key"),
        (re.compile(r"AKIA[0-9A-Z]{16}"), "AWS Access Key ID"),
        (re.compile(r"gh[pousr][_-][A-Za-z0-9_]{36,255}"), "GitHub Personal Access Token"),
        (re.compile(r"xox[baprs]-[0-9]{10,13}-[0-9]{10,13}-[a-zA-Z0-9]{24,32}"), "Slack Token"),
        (re.compile(r"-----BEGIN (?:RSA )?PRIVATE KEY-----"), "Private Cryptographic Key"),
    ]

    def evaluate(self, event: HookEvent) -> Optional[HookResult]:
        output = event.output
        if output is None:
            return None

        text = str(output)
        for pattern, label in self.PATTERNS:
            if pattern.search(text):
                return HookResult(
                    event_id=event.event_id,
                    verdict=HookVerdict.WARN,
                    message=f"SECRET LEAK DETECTED: {label} found in output stream. Redacting.",
                    exit_code=0,
                    rule_id=self.rule_id,
                    metadata={"leak_type": label},
                )
        return None


class CircuitBreakerRule(PolicyRule):
    """Trips and halts runaway agent loops when consecutive failure streaks >= 2."""
    rule_id = "RES-006-CIRCUIT-BREAKER"
    description = "Halts speculative retry loops when failure threshold is exceeded"

    def __init__(self, failure_threshold: int = 2):
        self.failure_threshold = failure_threshold
        self.streak_counters: Dict[str, int] = {}

    def record_failure(self, agent_id: str):
        self.streak_counters[agent_id] = self.streak_counters.get(agent_id, 0) + 1

    def record_success(self, agent_id: str):
        self.streak_counters[agent_id] = 0

    def evaluate(self, event: HookEvent) -> Optional[HookResult]:
        agent_id = event.agent_id or "default_agent"
        streak = self.streak_counters.get(agent_id, 0)
        if streak >= self.failure_threshold:
            return HookResult(
                event_id=event.event_id,
                verdict=HookVerdict.BLOCK,
                message=(
                    f"CIRCUIT BREAKER TRIPPED: Agent '{agent_id}' has {streak} consecutive failures. "
                    "Halting speculative edits. Run Abductive Diagnosis before retrying."
                ),
                exit_code=2,
                rule_id=self.rule_id,
            )
        return None


class UniversalPolicyEngine:
    """Orchestrates all policy rules and computes conclusive hook verdicts."""

    def __init__(self, project_dir: Optional[str] = None):
        self.project_dir = project_dir or str(Path.cwd())
        self.rules: List[PolicyRule] = [
            DestructiveCommandRule(),
            PackagePolicyRule(),
            SensitiveFileRule(),
            TddIntegrityRule(self.project_dir),
            SecretLeakRule(),
            CircuitBreakerRule(),
        ]

    def add_rule(self, rule: PolicyRule):
        self.rules.append(rule)

    def evaluate(self, event: HookEvent) -> HookResult:
        warnings = []
        for rule in self.rules:
            result = rule.evaluate(event)
            if result:
                if result.verdict == HookVerdict.BLOCK:
                    return result
                elif result.verdict == HookVerdict.WARN:
                    warnings.append(result.message)

        if warnings:
            return HookResult(
                event_id=event.event_id,
                verdict=HookVerdict.WARN,
                message="; ".join(warnings),
                exit_code=0,
            )

        return HookResult(
            event_id=event.event_id,
            verdict=HookVerdict.ALLOW,
            message="Operation allowed by policy",
            exit_code=0,
        )

File: src/hooks/policy-engine.ts
/**
 * Universal Agentic Hooks Framework (UAHF) — Policy Engine (TypeScript/Bun)
 * =========================================================================
 * Deterministic policy evaluator matching Python engine rules with sub-millisecond execution.
 */

import fs from 'node:fs';
import path from 'node:path';
import type { HookEvent, HookResult, PolicyRule } from './types.js';

export class DestructiveCommandRule implements PolicyRule {
  rule_id = 'SEC-001-DESTRUCTIVE-COMMAND';
  description = 'Blocks destructive system, git, and exfiltration commands';

  private networkPatterns = [
    { regex: /\bcurl\b.*\|\s*(ba)?sh\b/, msg: 'curl piped to shell is blocked. Download, inspect, and execute manually.' },
    { regex: /\bwget\b.*\|\s*(ba)?sh\b/, msg: 'wget piped to shell is blocked. Download, inspect, and execute manually.' },
  ];

  private systemPatterns = [
    { regex: /\bdd\b\s+if=/, msg: 'dd command with raw input file is blocked. Irreversible disk write risk.' },
    { regex: /\bmkfs\b/, msg: 'mkfs command is blocked. Filesystem formatting destroys all data.' },
  ];

  private gitPatterns = [
    { regex: /\bgit\s+push\b.*(?<![-\w])--force(?![-\w])/, msg: 'git push --force is blocked. Use --force-with-lease to protect remote history.' },
    { regex: /\bgit\s+push\b.*\s-[a-zA-Z]*f/, msg: 'git push -f is blocked. Use --force-with-lease to protect remote history.' },
    { regex: /\bgit\s+reset\b.*\s--hard\b/, msg: 'git reset --hard is blocked. Discards uncommitted work permanently.' },
    { regex: /\bgit\s+checkout\b\s+(?:--\s+)?\./, msg: 'git checkout . is blocked. Discards unstaged modifications.' },
    { regex: /\bgit\s+restore\b\s+\./, msg: 'git restore . is blocked. Discards unstaged modifications.' },
    { regex: /\bgit\s+clean\b.*\s-[a-zA-Z]*f/, msg: 'git clean -f is blocked. Permanently deletes untracked files.' },
    { regex: /\bgit\s+branch\b.*\s-D\b/, msg: 'git branch -D is blocked. Use git branch -d for safe deletion.' },
    { regex: /\bgit\s+branch\b.*\s--delete\b.*\s--force\b/, msg: 'git branch --delete --force is blocked. Use git branch -d for safe deletion.' },
  ];

  private dangerousTargets = new Set(['/', '/*', '~', '~/', '$HOME', '$HOME/', '$HOME/*']);

  private checkDangerousRm(subCommand: string): string | null {
    const tokens = subCommand.trim().split(/\s+/);
    if (!tokens.length || tokens[0] !== 'rm') return null;

    let flags = '';
    const targets: string[] = [];

    for (let i = 1; i < tokens.length; i++) {
      const t = tokens[i];
      if (t.startsWith('-') && !t.startsWith('--')) {
        flags += t.slice(1);
      } else if (!t.startsWith('-')) {
        targets.push(t.replace(/^['"]|['"]$/g, ''));
      }
    }

    const hasRecursive = flags.includes('r') || flags.includes('R');
    const hasForce = flags.includes('f');

    if (!hasRecursive || !hasForce) return null;

    for (const tgt of targets) {
      if (this.dangerousTargets.has(tgt)) {
        return `rm -rf targeting ${tgt} is blocked. Catastrophic, irreversible file deletion.`;
      }
    }
    return null;
  }

  evaluate(event: HookEvent): HookResult | null {
    const cmd = event.command || (event.args?.command as string);
    if (!cmd || typeof cmd !== 'string') return null;

    for (const { regex, msg } of this.networkPatterns) {
      if (regex.test(cmd)) {
        return {
          event_id: event.event_id,
          verdict: 'block',
          message: `DESTRUCTIVE COMMAND BLOCKED: ${msg}`,
          exit_code: 2,
          rule_id: this.rule_id,
        };
      }
    }

    for (const { regex, msg } of this.systemPatterns) {
      if (regex.test(cmd)) {
        return {
          event_id: event.event_id,
          verdict: 'block',
          message: `DESTRUCTIVE COMMAND BLOCKED: ${msg}`,
          exit_code: 2,
          rule_id: this.rule_id,
        };
      }
    }

    for (const { regex, msg } of this.gitPatterns) {
      if (regex.test(cmd)) {
        return {
          event_id: event.event_id,
          verdict: 'block',
          message: `DESTRUCTIVE COMMAND BLOCKED: ${msg}`,
          exit_code: 2,
          rule_id: this.rule_id,
        };
      }
    }

    const subCommands = cmd.split(/\s*(?:&&|\|\||;)\s*/);
    for (const sc of subCommands) {
      for (const seg of sc.split('|')) {
        const rmErr = this.checkDangerousRm(seg);
        if (rmErr) {
          return {
            event_id: event.event_id,
            verdict: 'block',
            message: `DESTRUCTIVE COMMAND BLOCKED: ${rmErr}`,
            exit_code: 2,
            rule_id: this.rule_id,
          };
        }
      }
    }

    return null;
  }
}

export class PackagePolicyRule implements PolicyRule {
  rule_id = 'PKG-002-PACKAGE-HYGIENE';
  description = 'Enforces container and package manager governance policies';

  private forbiddenPackages: Record<string, string> = {
    colima: 'Colima is prohibited per project constitution due to large footprint. Use lightweight alternatives.',
  };

  private brewInstallRegex = /\bbrew\s+(?:install|reinstall|cask)\s+([^\s;]+)/i;
  private sudoBrewRegex = /\bsudo\s+brew\b/i;

  evaluate(event: HookEvent): HookResult | null {
    const cmd = event.command || (event.args?.command as string);
    const argsTarget = (event.args?.target || event.args?.package) as string;

    if (cmd && typeof cmd === 'string') {
      if (this.sudoBrewRegex.test(cmd)) {
        return {
          event_id: event.event_id,
          verdict: 'block',
          message: "PACKAGE POLICY VIOLATION: 'sudo brew' is prohibited. Homebrew must not run as root.",
          exit_code: 2,
          rule_id: this.rule_id,
        };
      }

      const match = cmd.match(this.brewInstallRegex);
      if (match) {
        const pkgCandidate = match[1].toLowerCase().replace(/^['"]|['"]$/g, '');
        for (const [forbidden, reason] of Object.entries(this.forbiddenPackages)) {
          if (pkgCandidate.includes(forbidden)) {
            return {
              event_id: event.event_id,
              verdict: 'block',
              message: `PACKAGE POLICY VIOLATION: Package '${forbidden}' is blocked. ${reason}`,
              exit_code: 2,
              rule_id: this.rule_id,
            };
          }
        }
      }
    }

    if (argsTarget && typeof argsTarget === 'string') {
      const tgt = argsTarget.toLowerCase();
      for (const [forbidden, reason] of Object.entries(this.forbiddenPackages)) {
        if (tgt === forbidden || tgt.includes(forbidden)) {
          return {
            event_id: event.event_id,
            verdict: 'block',
            message: `PACKAGE POLICY VIOLATION: Package '${forbidden}' is blocked. ${reason}`,
            exit_code: 2,
            rule_id: this.rule_id,
          };
        }
      }
    }

    return null;
  }
}

export class SensitiveFileRule implements PolicyRule {
  rule_id = 'SEC-003-SENSITIVE-FILES';
  description = 'Prevents tampering with credentials, environment secrets, and private keys';

  private sensitivePatterns = [
    { regex: /(^|[/\\])\.env(\.[a-zA-Z0-9_-]+)?$/, label: 'Environment secret file (.env)' },
    { regex: /\.(pem|key|p12|pfx)$/i, label: 'Private key or certificate' },
    { regex: /(^|[/\\])id_(rsa|ed25519|ecdsa|dsa)$/, label: 'SSH private key' },
    { regex: /(^|[/\\])(credentials|secrets|passwords)\.(json|ya?ml|toml)$/i, label: 'Secrets store' },
    { regex: /(^|[/\\])(service[-_]?account|token)\.json$/i, label: 'Service account / Token file' },
    { regex: /\.(tfstate|tfvars)$/i, label: 'Terraform state / variable secrets' },
  ];

  evaluate(event: HookEvent): HookResult | null {
    const filePath = event.file_path || (event.args?.file_path as string) || (event.args?.path as string);
    if (!filePath || typeof filePath !== 'string') return null;

    for (const { regex, label } of this.sensitivePatterns) {
      if (regex.test(filePath)) {
        return {
          event_id: event.event_id,
          verdict: 'warn',
          message: `SECURITY ADVISORY: Access/modification to sensitive file '${filePath}' (${label}).`,
          exit_code: 0,
          rule_id: this.rule_id,
        };
      }
    }
    return null;
  }
}

export class TddIntegrityRule implements PolicyRule {
  rule_id = 'GOV-004-TDD-INTEGRITY';
  description = 'Protects test files from modification during implementation phases';

  private testFilePatterns = [
    /(^|[/\\])test_[^/\\]+\.py$/,
    /[._]test\.[jt]sx?$/,
    /[._]spec\.[jt]sx?$/,
  ];

  constructor(private projectDir: string = process.cwd()) {}

  isGuardActive(): boolean {
    return (
      fs.existsSync(path.join(this.projectDir, '.tdd-guard')) ||
      fs.existsSync(path.join(process.cwd(), '.tdd-guard'))
    );
  }

  evaluate(event: HookEvent): HookResult | null {
    if (!this.isGuardActive()) return null;

    const filePath = event.file_path || (event.args?.file_path as string) || (event.args?.path as string);
    if (!filePath || typeof filePath !== 'string') return null;

    const tool = (event.tool_name || '').toLowerCase();
    if (!tool.includes('edit') && !tool.includes('write')) return null;

    for (const pattern of this.testFilePatterns) {
      if (pattern.test(filePath)) {
        return {
          event_id: event.event_id,
          verdict: 'block',
          message: `TDD GUARD ACTIVE: Modification of test file '${filePath}' is blocked. Modify implementation code to make tests pass.`,
          exit_code: 2,
          rule_id: this.rule_id,
        };
      }
    }
    return null;
  }
}

export class SecretLeakRule implements PolicyRule {
  rule_id = 'SEC-005-SECRET-LEAK-FILTER';
  description = 'Detects and blocks leakage of API keys and authentication tokens in telemetry';

  private patterns = [
    { regex: /sk-(?:proj|ant|live)-[a-zA-Z0-9_\-]{20,}/, label: 'OpenAI / Anthropic API Key' },
    { regex: /AKIA[0-9A-Z]{16}/, label: 'AWS Access Key ID' },
    { regex: /gh[pousr][_-][A-Za-z0-9_]{36,255}/, label: 'GitHub Personal Access Token' },
    { regex: /xox[baprs]-[0-9]{10,13}-[0-9]{10,13}-[a-zA-Z0-9]{24,32}/, label: 'Slack Token' },
    { regex: /-----BEGIN (?:RSA )?PRIVATE KEY-----/, label: 'Private Cryptographic Key' },
  ];

  evaluate(event: HookEvent): HookResult | null {
    if (!event.output) return null;
    const text = typeof event.output === 'string' ? event.output : JSON.stringify(event.output);

    for (const { regex, label } of this.patterns) {
      if (regex.test(text)) {
        return {
          event_id: event.event_id,
          verdict: 'warn',
          message: `SECRET LEAK DETECTED: ${label} found in output stream. Redacting.`,
          exit_code: 0,
          rule_id: this.rule_id,
          metadata: { leak_type: label },
        };
      }
    }
    return null;
  }
}

export class CircuitBreakerRule implements PolicyRule {
  rule_id = 'RES-006-CIRCUIT-BREAKER';
  description = 'Halts speculative retry loops when failure threshold is exceeded';

  private streakCounters: Map<string, number> = new Map();

  constructor(private failureThreshold: number = 2) {}

  recordFailure(agentId: string) {
    const current = this.streakCounters.get(agentId) || 0;
    this.streakCounters.set(agentId, current + 1);
  }

  recordSuccess(agentId: string) {
    this.streakCounters.set(agentId, 0);
  }

  evaluate(event: HookEvent): HookResult | null {
    const agentId = event.agent_id || 'default_agent';
    const streak = this.streakCounters.get(agentId) || 0;
    if (streak >= this.failureThreshold) {
      return {
        event_id: event.event_id,
        verdict: 'block',
        message: `CIRCUIT BREAKER TRIPPED: Agent '${agentId}' has ${streak} consecutive failures. Halting speculative edits. Run Abductive Diagnosis before retrying.`,
        exit_code: 2,
        rule_id: this.rule_id,
      };
    }
    return null;
  }
}

export class UniversalPolicyEngine {
  public rules: PolicyRule[];

  constructor(public projectDir: string = process.cwd()) {
    this.rules = [
      new DestructiveCommandRule(),
      new PackagePolicyRule(),
      new SensitiveFileRule(),
      new TddIntegrityRule(this.projectDir),
      new SecretLeakRule(),
      new CircuitBreakerRule(),
    ];
  }

  addRule(rule: PolicyRule) {
    this.rules.push(rule);
  }

  evaluate(event: HookEvent): HookResult {
    const warnings: string[] = [];
    for (const rule of this.rules) {
      const res = rule.evaluate(event);
      if (res) {
        if (res.verdict === 'block') {
          return res;
        } else if (res.verdict === 'warn') {
          warnings.push(res.message);
        }
      }
    }

    if (warnings.length > 0) {
      return {
        event_id: event.event_id,
        verdict: 'warn',
        message: warnings.join('; '),
        exit_code: 0,
      };
    }

    return {
      event_id: event.event_id,
      verdict: 'allow',
      message: 'Operation allowed by policy',
      exit_code: 0,
    };
  }
}
